src/main/java/org/wildfly/security/x500/cert/CertUtil.java

```diff
@@ -0,0 +1,380 @@
+    private static final String IP = "IP";
+    private static final String OID = "OID";
+    private static final String[] ALT_NAMES_TYPES = new String[] { EMAIL, URI, DNS, IP, OID };
+    private static final int[] DELIMS = new int[] {',', ' '};
+
+    /**
+     * Get the default compatible signature algorithm name for the given private key.
+     *
+     * @param privateKey the private key
+     * @return the default compatible signature algorithm name for the given private key or {@code null}
+     * if the default compatible signature algorithm name cannot not be determined
+     * @throws IllegalArgumentException if the key size cannot be determined from the given private key
+     */
+    public static String getDefaultCompatibleSignatureAlgorithmName(final PrivateKey privateKey) throws IllegalArgumentException {
+        final int keySize = getKeySize(privateKey);
+        if (keySize == -1) {
+            throw log.unableToDetermineKeySize();
+        }
+        return getDefaultCompatibleSignatureAlgorithmName(privateKey.getAlgorithm(), keySize);
+    }
+
+    /**
+     * Get the default compatible signature algorithm name for the given key algorithm name and key size.
+     *
+     * @param keyAlgorithmName the key algorithm name
+     * @param keySize the key size
+     * @return the default compatible signature algorithm name for the given key algorithm name and key size
+     * or {@code null} if the default compatible signature algorithm name cannot not be determined
+     */
+    public static String getDefaultCompatibleSignatureAlgorithmName(final String keyAlgorithmName, final int keySize) {
+        final String messageDigestAlgorithmName = getDefaultMessageDigestAlgorithmName(keyAlgorithmName, keySize);
+        if (messageDigestAlgorithmName == null) {
+            return null;
+        }
+        switch (keyAlgorithmName) {
+            case "DSA": {
+                return messageDigestAlgorithmName + "withDSA";
+            }
+            case "RSA": {
+                return messageDigestAlgorithmName + "withRSA";
+            }
+            case "EC": {
+                return messageDigestAlgorithmName + "withECDSA";
+            }
+            default: {
+                return null;
+            }
+        }
+    }
+
+    /**
+     * Get the default message digest algorithm name for the given key algorithm name and key size.
+     *
+     * @param keyAlgorithmName the key algorithm name
+     * @param keySize the key size
+     * @return the default message digest algorithm name or {@code null} if the default message algorithm name cannot be determined
+     */
+    private static String getDefaultMessageDigestAlgorithmName(final String keyAlgorithmName, final int keySize) {
+        // These defaults are based on Tables 2 and 3 in NIST SP 800-57 Part 1 Revision 4
+        // (see https://csrc.nist.gov/publications/detail/sp/800-57-part-1/rev-4/final)
+        switch (keyAlgorithmName) {
```

Comment:
Is this likely to be something we need to keep up to date?

Reply:
This is something that could change but it looks like it doesn't happen very often. NIST SP 800-57 Part 1 Revision 4 was published in 2016 but the previous revision was published in 2012.